### PR TITLE
Introduce MPI allreduce and allgather in a new contrib project

### PR DESCRIPTION
**Pull request by jthestness**
This commit adds the tensorflow.contrib.mpi_collectives namespace and
contrib project, which has a variety of ops that work with MPI.

The MPI system works by starting a background thread which communicates
between the different processes at a regular interval and schedules
asynchronous reductions. At every tick, every rank will notify rank zero
of the tensors it is ready to reduce, signifying completion with an
empty DONE message. Rank zero will count how many ranks are ready to
reduce every tensor, and, whenever a tensor is ready to reduce (that is,
every rank is ready to reduce it), rank zero will issue a message to all
other ranks directing them to reduce that tensor.  This repeats for all
the tensors that are ready to reduce, after which rank zero sends all
other ranks a DONE message indicating that the tick is complete.

Authors: Andrew Gibiansky <andrew.gibiansky@gmail.com>
               Joel Hestness <jthestness@gmail.com>

### CODE REVIEW

**Comment by jthestness** [comment]
PR submitted from Baidu Silicon Valley AI Lab (SVAIL)

**Comment by alsrgv** [comment]
@jthestness FYI - we made derivative work based on your original https://github.com/baidu-research/tensorflow-allreduce called Horovod: https://github.com/uber/horovod.  It's a plugin that installs on top of TensorFlow, so you can `pip install tensorflow` and then `pip install horovod`, no need to Bazel-build anything.

We made number of usability improvements and integrated NCCL 2 support, which makes it easier to use and it performs faster.  We've also added special Horovod Timeline that aids debugging of overall distributed training.  Now we're working on further optimisation techniques to gain further speed-up on networks with many tiny elements like `resnet152` on large # of machines.

**Review by poxvoculi — APPROVED** [review]
Sorry this took so long; I was on vacation.   Overall this looks really good.  I've got a few minor observations, but should be mergeable as it.

**Comment by byronyi** [comment]
Nice work!

Sorry to be off-topic, but what happened to @gibiansky? Is he still in Baidu SVAIL?

**Comment by gibiansky** [comment]
@byronyi I am still here! I've been working with @jthestness to get this together (well, mostly @jthestness has been working, and I've been offering the occasional useful advice :P). (Hasn't been my primary project as I got distracted with other things while we were working on getting the CLA signed.)

File: tensorflow/contrib/mpi_collectives/mpi_allreduce_test.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    my_rank = int(os.environ['PMI_RANK'])
+    num_ranks = int(os.environ['PMI_SIZE'])
+
+    stages = 13
+    batch_size = 1331
+    hidden_size = batch_size
+    out_size = batch_size
+
+    # Input placeholder (batch_size x hidden) - init to 1s
+    inputs = tf.placeholder(tf.float32, shape=(batch_size, hidden_size),
```
[the remark this answers is not in the document]
Comment (jthestness) [reply tensorflow/contrib/mpi_collectives/mpi_allreduce_test.py]: We have only used internal RNN models to verify correctness of the allgather on IndexedSlices rather than unit tests. I'll create a test if you prefer.

File: tensorflow/contrib/mpi_collectives/__init__.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+  This function performs a bandwidth-optimal ring allreduce on the input
+  tensor. If the input is an tf.IndexedSlices, the function instead does an
+  allgather on the values and the indices, effectively doing an allreduce on
+  the represented tensor.
+  """
+  if isinstance(tensor, tf.IndexedSlices):
+    # For IndexedSlices, do two allgathers intead of an allreduce.
+    mpi_size = tf.cast(size(), tensor.values.dtype)
+    values = allgather(tensor.values)
```
[the remark this answers is not in the document]
Comment (jthestness) [reply tensorflow/contrib/mpi_collectives/__init__.py]: IndexedSlices tensors are allowed to have repeated indices, so gathering and concatting IndexedSlices together without any reduction or index merging still creates a valid IndexedSlices tensor. For weight updates during backprop, a gradient tensor can be IndexedSlices with repeated indices, so we can call `apply_gradients` with such a tensor.

There is a question about whether this organization is an appropriate design for the interface. A user might (reasonably) assume that the returned IndexedSlices tensor should have unique indices.

I'm happy to discuss appropriate organization, and as necessary, adjust the code or add comments.

[the remark this answers is not in the document]
Comment (poxvoculi) [reply tensorflow/contrib/mpi_collectives/__init__.py]: I wasn't familiar with IndexedSlices; if repeated indices are ok, then this is fine.

File: tensorflow/contrib/mpi_collectives/mpi_allreduce_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (poxvoculi) [reply tensorflow/contrib/mpi_collectives/mpi_allreduce_test.py]: Up to you, but at least one small test seems appropriate.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by gibiansky** [comment]
Are the failing tests issues with this PR? I took a look at the console logs from them, and couldn't find anything that looked like a failure due to the MPI allreduce.

**Comment by martinwicke** [comment]
Jenkins, test this please.

File: tensorflow/contrib/mpi_collectives/mpi_allreduce_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jthestness) [reply tensorflow/contrib/mpi_collectives/mpi_allreduce_test.py]: Added an Allgather test that runs correctly on our systems

File: tensorflow/contrib/mpi_collectives/mpi_message.proto [hunk 3]
```diff
@@ -0,0 +1,63 @@
+syntax = "proto3";
+
+package tensorflow.contrib.mpi;
+
+// We would like to just use DataType here, but since this
+// is a contrib package, linking directly to TensorFlow protos seems to be
+// impossible. Doing so compiles, but fails with a cryptic error at runtime
+// about a pointer that was passed to free() but not created by malloc().
```
[the remark this answers is not in the document]
Comment (jthestness) [reply tensorflow/contrib/mpi_collectives/mpi_message.proto]: I was able to fix the linking against other protos, so I removed the MPIDataType in favor of the TF DataType proto.

**Comment by jthestness** [comment]
Just wanted to check in here. Is this PR ready to merge or do we need anything else from my side?

**Comment by martinwicke** [comment]
Jenkins, test this please.

@gunan this PR didn't show the mandatory checks -- did you change the name and I didn't see it or did the tests run but didn't report status?

**Comment by martinwicke** [comment]
Hmm, not it seems to work. Eh. I blame Github.

**Comment by martinwicke** [comment]
There are build errors, likely because of mpi not being appropriately declared in the workspace.

**Comment by gunan** [comment]
It is possible I made the rename and this PR did not get updated. after it. or something like that.
It may have stayed stale until we reran the tests.

**Comment by jthestness** [comment]
Cool. Will try to fix these ASAP.

**Comment by gunan** [comment]
That is actually already fixed. All the failures you see now are related to this PR:
```
In file included from tensorflow/contrib/mpi_collectives/ring.cu.cc:6:0:
./tensorflow/contrib/mpi_collectives/ring.h:17:33: fatal error: third_party/mpi/mpi.h: No such file or directory
 #include "third_party/mpi/mpi.h"
```

Some BUILD dependencies are missing.

**Comment by chenliu0831** [comment]
Just curious can we provide an example (e.g. mnist) like [here](https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py)

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by jthestness** [comment]
Looks like one of the GPU tests timed out. Does this need to be re-run or was the successful build sufficient?

**Comment by drpngx** [comment]
Yeah, `fft_ops_test` will be fixed in the next push.

**Comment by drpngx** [comment]
Ignoring `fft_ops_test` failure.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Windows build appears to have come back online.

Jenkins, test this please.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Something wrong with the CLA bot. Let's wait a bit.

**Comment by drpngx** [comment]
OK, now the CLA looks good. 

Jenkins, test this please.

**Comment by drpngx** [comment]
Jenkins, test this please. (stuck)

**Comment by jhseu** [comment]
Thanks Joel for the amazing contribution!

**Comment by shamoya** [comment]
Very nice.
Do we have a modified version of [benchmarks](https://github.com/tensorflow/benchmarks) (or any other 
 popular models benchmark) modified to use this API ?

**Comment by alexwal** [comment]
This is great, thanks for the contribution. I was wondering, what is your take on https://github.com/uber/horovod?

**Comment by Zhaojp-Frank** [comment]
Impressive work! A few questions to setup my env, pls kindly help: thanks a lot.
1. Regarding MPI dependency, is it always required in both inter-node and intra-node cases? such as a single node with multi-GPU, is MPI still required? if inter-node, what's the relationship btw gRPC and MPI?
 not sure, any other possible choice if I want to remove MPI dependency?
2. Does it support Nvidia NCCL? if so, is NCCL internally handled by MPI?

**Comment by jthestness** [comment]
@alexwal : We haven't tested Horovod, but the Uber team claims it performs better than this MPI allreduce. We have recently tested NCCL 2.0 (i.e. what Horovod uses) for intra-node communication outside of Tensorflow, and it performs a bit better than the MPI ring allreduce. We have some analysis about the differences that we will be releasing in the near future.

@Zhaojp-Frank : (1) Tensorflow supports intra-node communication, and you can use that for data parallelism if you'd like. It cannot currently scale to inter-node communication without a collectives package such as this MPI collectives. Tensorflow's intra-node communication requires that you use a single process to manage all GPUs, while inter-node communication requires at least one process on each separate node. Because of these differences, the programming model can be very different. If you plan to scale to multiple nodes, we recommend that you start with collectives that will scale to multiple nodes rather than trying to retrofit that code later. (2) This MPI collectives package doesn't use NCCL.

**Comment by Zhaojp-Frank** [comment]
@jthestness thanks a lot.  does the implementation also overlap AllReduce communication and next layer gradient computing?
beside NCCL support, Horovod also batch small tensors during communication, that may also help performance.

**Comment by camaclean** [comment]
Why is an allreduce manually implemented using point to point communications rather than using the likely much more optimized allreduce present in MPI? Do you have a specific reason for doing it yourself?

Vendor MPI implementations like cray-mpich can use hardware support for collective communications and take advantage of tightly interconnected networks and don't load a single node with scheduling concerns. A ring allreduce has an abysmal performance in comparison on such networks, especially at scale (thousands of nodes).

It would also take a lot less code to just `MPI_Iallreduce` when a rank has a tensor ready to reduce, then either periodically use `MPI_Testall` on waiting reduces or have a background thread use `MPI_Waitany` to progress the allreduce. Is there a reason why you can't just do this?

**Comment by gibiansky** [comment]
@camaclean 

Short version: The allreduce present in MPI -- that is, non-vendor MPI implementations -- is not CUDA-aware, and thus copies memory to CPU before doing the collective. The ring allreduce is much more efficient when doing multi-node multi-GPU training than the default collective.

When writing this package, we were optimizing for commodity hardware (possibly with Infiniband for multi-node connections) on CPUs and GPUs, which describes the majority of large DL applications. At the time of writing, most MPI implementations (that is, all that I checked) did not support direct GPU-to-GPU transfers in their collectives; however, at least Open MPI supported the simpler operations using direct GPU-to-GPU transfers and CUDA IPC. If you used `MPI_Iallreduce`, it would first transfer GPU memory to CPU, then perform the allreduce, then transfer it back. For a model with 1 GB of weights with several GPUs connected to the CPU via a shared PCIe connection, this would introduce a significant performance penalty. Implementing our own ring allreduce allowed us to take advantage of GPU-to-GPU transfers without crossing into CPU memory, so that the slowest point was (on multi-node setups) the network or Infiniband connection. The difference in performance between using `MPI_Iallreduce` and our ring allreduce was easily an order of magnitude.

However, as you mention, this code is certainly not optimal for all MPI implementations and network topologies. Additionally, nowadays NCCL supports (or claims to support) multi-node communication, although I haven't had the time to check how well it works or how it performs its collectives (and whether they correctly handle GPU to GPU transfers over infiniband).

If you have access to a thousand node cluster and want to switch out the allreduce for the default vendor one, would be really cool to see what sort of things you could do :)

**Comment by poxvoculi** [comment]
FYI:  core TensorFlow now has collective all-reduce and broadcast, though not yet part of the official API.  

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/collective_ops.py

The strategic intention is that users should use the new distribution strategies framework to specify data-parallel distribution, and that framework will utilize collectives as appropriate.  

https://www.tensorflow.org/api_docs/python/tf/contrib/distribute/DistributionStrategy

The default implementation of collectives uses RPCs over gRPC so the performance is not great, compared to MPI/ib.  However, MPI is not available everywhere.  Going forward we plan to investigate introducing an API that makes it possible to link in MPI collectives where available and get platform-customized implementations, e.g. for a supercomputer.

**Comment by camaclean** [comment]
@gibiansky 

Ah, I see. That makes sense. Unfortunately, our hardware is old and doesn't support true cuda aware MPI. We have 4,228 GPU nodes, each with a Tesla K20X, networked in a 3d torus. Each node isn't very fast, but we make up for it by having a lot of them (there's an additional 22,636 CPU nodes, but those aren't too useful for ML). It means having applications that can scale well, though, and getting Tensorflow to do so well has been painful. Cray has a Tensorflow ML plugin written specifically for their interconnects, but it's a bit finicky on our system and they don't really want to invest much effort in our older model system (It makes sense, but it's a struggle for me). So, I'm constantly on the look out for things to try and make Tensorflow scale better. We only recently upgraded from CUDA 7.5 to CUDA 9.1, so we were stuck on Tensorflow 1.4.1 for a long time, but now I'm back to investigating it.

Have you tried doing a local intermediate reduction within the gpus on each node, then ireducing that? We only have one gpu per node, so it's not something I can test. Perhaps I'll have to write an HPC-orriented plugin of my own, since most are one gpu per node with larger node counts as opposed to high-gpu-count ML clusters like what you're targeting. Or perhaps there could be choices between allreduce methods to help support a broader range of hardware configurations.

@poxvoculi 

Good to know. I just hope the CUDA requirements don't evolve too fast to prevent us from upgrading when that feature is ready. As I mentioned above, we only just recently moved from CUDA 7.5 to CUDA 9.1, and CUDA 9.1 is all we will get until the end of the machine's life (probably the end of 2019). HPC system software tends to be updated more slowly than Tensorflow's dependencies since everything needs to be well tested, which I know has been an issue for other centers, as well. A lot of our users want the latest and greatest Tensorflow, but actually getting it running and performing well was the topic of a number of papers at this year's Cray User Group conference. I know HPC is only a small fraction of total users, but I'm glad you're adjusting the API to make it easier to run well on our big systems.

**Comment by gibiansky** [comment]
@camaclean That's really cool! That said, getting a model to train well on that giant set of nodes seems like a pretty big challenge... With regards to the intermediate reduction: since we were operating within a single infiniband switch, a ring allreduce meant that we were just bandwidth limited between adjacent nodes in the allreduce. We would do 8-10 GPUs on a single machine and then have up to 10-30 nodes at most. So the only thing we tried besides the ring was a few special cases for when reducing small tensors (where latency dominates bandwidth). 

(Getting an efficient synchronous gradient update on a setup that big seems very tricky. Additionally, even if you do, your batch size will be *enormous*... if you're training standard deep learning models you'll have to throw a bunch of tricks at it to get it to optimize well with that many elements in the batch. Maybe a (partially?) asynchronous training strategy is more feasible?)

**Comment by camaclean** [comment]
@gibiansky 

For some reason, getting things to scale on Blue Waters has been harder than what I've seen from other centers. Last time we were benchmarking we were having trouble with crashes and such. Pytorch was much easier to get scaling well since it can use MPI's allreduce under the hood. Our network is a 3d torus with 2 nodes sharing an interconnect, which can do 2.9GB/s (effective) for each of the 6 directions and intelligently route packets along several links if there's bandwidth available. It also has hardware support for doing collective type operations to efficiently scatter data. Based on profiling counters on the interconnects, we haven't been able to come close to saturating our network bandwidth. Our biggest bottleneck seems to be more a matter of latency.

**Comment by snowpeakz** [comment]
Is there a complete training code example with ring allreduce implemented or a document about refactoring code with ring allreduce?

**Comment by gibiansky** [comment]
@snowpeakz At this point, I would recommend using [Horovod](https://github.com/horovod/horovod), which is based on this code but improves it and has pretty good documentation.

**Comment by snowpeakz** [comment]
@gibiansky thanks for recommendation

**Comment by lxl910915** [comment]
@gibiansky If the number of worker is large than 12, the cmd "mpiexec -n 16 --hostfile /tmp/hostfile --allow-run-as-root -q sh -c 'echo $(hostname) is ready' " return with error code 205 ?
Any suggestion? Thanks.

**Comment by gibiansky** [comment]
@snowpeakz no idea I'm afraid.

**Comment by lxl910915** [comment]
> @snowpeakz no idea I'm afraid.

Do you use mpiexec to run your code?
